substrate/frame/nis/src/migration.rs

```diff
@@ -0,0 +1,261 @@
+}
+
+pub mod switch_block_number_provider {
+	use super::*;
+
+	/// The log target.
+	const TARGET: &'static str = "runtime::nis::migration::change_block_number_provider";
+
+	pub trait BlockNumberConversion<Old, New> {
+		fn convert_block_number(block_number: Old) -> New;
+	}
+
+	pub struct MigrateBlockNumber<T, BlockConverter>(PhantomData<T>, PhantomData<BlockConverter>);
+	impl<
+			T: Config,
+			BlockConverter: BlockNumberConversion<SystemBlockNumberFor<T>, BlockNumberFor<T>>,
+		> OnRuntimeUpgrade for MigrateBlockNumber<T, BlockConverter>
+	{
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, TryRuntimeError> {
+			let old_receipts = v0::Receipts::<T>::iter().collect::<Vec<_>>();
+			let old_summary = v0::Summary::<T>::get();
+			Ok((old_receipts, old_summary).encode())
+		}
+
+		fn on_runtime_upgrade() -> Weight {
+			let mut weight = Weight::zero();
+			weight.saturating_accrue(migrate_block_number::<T, BlockConverter>());
+			weight
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(state: Vec<u8>) -> Result<(), TryRuntimeError> {
+			// Decode pre-upgrade state
+			let (old_receipts, old_summary): (
+				Vec<(ReceiptIndex, v0::OldReceiptRecordOf<T>)>,
+				v0::OldSummaryRecordOf<T>,
+			) = Decode::decode(&mut &state[..]).expect("pre_upgrade data must decode");
+
+			// Assert the count of receipts after migration matches the original count
+			ensure!(
+				Receipts::<T>::iter().count() == old_receipts.len(),
+				"Receipt count mismatch after migration"
+			);
+
+			// Verify Receipts migration
+			for (index, old_receipt) in old_receipts {
+				let expected_expiry = BlockConverter::convert_block_number(old_receipt.expiry);
+				let new_receipt =
+					Receipts::<T>::get(index).ok_or("Receipt missing after migration")?;
+
+				// Verify expiry conversion
+				ensure!(new_receipt.expiry == expected_expiry, "Receipt expiry conversion failed");
+
+				// Verify other fields unchanged
+				ensure!(
+					new_receipt.proportion == old_receipt.proportion &&
+						new_receipt.owner == old_receipt.owner,
+					"Receipt fields corrupted"
+				);
+			}
+
+			// Verify Summary migration
+			let new_summary = Summary::<T>::get();
+			let expected_last_period =
+				BlockConverter::convert_block_number(old_summary.last_period);
+			ensure!(new_summary.last_period == expected_last_period, "Summary conversion failed");
+			log::info!(target: TARGET, "All reciept record expiry period and summary record thaw period begining migrated to new blok number provider");
+			Ok(())
+		}
+	}
+
+	pub fn migrate_block_number<T, BlockConverter>() -> Weight
+	where
+		BlockConverter: BlockNumberConversion<SystemBlockNumberFor<T>, BlockNumberFor<T>>,
+		T: Config,
+	{
+		let on_chain_version = Pallet::<T>::on_chain_storage_version();
+		let mut weight = T::DbWeight::get().reads(1);
+		log::info!(
+			target: TARGET,
+			"running migration with onchain storage version {:?}", on_chain_version
+		);
+
+		if on_chain_version == 0 {
```

Comment:
Does not needs to be bound to the storage version. Also later you could do a change of the block number source.